### ISSUE TITLE
Building a standalone C++ Tensorflow program on Windows

### ISSUE THREAD

**Issue by joe-antognini** [joe-antognini/joe-antognini.github.io#9]
Written on 06/22/2017 00:37:00 

URL: https://joe-antognini.github.io/machine-learning/windows-tf-project

**Comment by joe-antognini** [comment]
Comment written by **AlexPetrenko** on **07/11/2017 18:08:39**

<p>I've also just managed to build tensorflow.dll shared library using VS2017. In order to do this I had to add a few cmake options to tensorflow build: `cmake .. -G"Visual Studio 15 2017 Win64" -Dtensorflow_BUILD_PYTHON_BINDINGS=OFF -Dtensorflow_ENABLE_GRPC_SUPPORT=OFF -Dtensorflow_BUILD_SHARED_LIB=ON`</p><p>Also I had to build protobuf that tensorflow pulls to cmake/build/protobuf directory separately with custom CMake flags:<br>`cmake .. -G"Visual Studio 15 2017 Win64" -DBUILD_SHARED_LIBS=ON -Dprotobuf_BUILD_TESTS=OFF` - this builds libprotobuf.dll, and I had to add this as a dependency for my project. Do not use latest libprotobuf release (3.3 something), it'll give you linkage errors.</p><p>Final command line for my own CMake project looked like this: cmake .. -G"Visual Studio 15 2017 Win64" -DTENSORFLOW_SRC_DIR="C:\all\lib\tensorflow-1.2.1" -DTENSORFLOW_BUILD_DIR="C:\all\lib\tensorflow-1.2.1\tensorflow\contrib\cmake\build" -DPROTOBUF_BUILD_DIR="C:\all\lib\tensorflow-1.2.1\tensorflow\contrib\cmake\build\protobuf\src\protobuf\cmake\build"`</p><p>After all that I managed to compile and run your matmul.cpp example! So happy :)<br>Sorry for a messy comment, hope this can help someone.</p>

**Comment by joe-antognini** [comment]
Comment written by **lotfi** on **07/31/2017 13:22:08**

<p>is there a way to Build a standalone C++ Tensorflow program on linux ?</p>

**Comment by joe-antognini** [comment]
Comment written by **antognini** on **07/31/2017 15:54:01**

<p>It should be easiest to use Bazel to build C++ Tensorflow on Linux.  I haven't done it myself yet, but this blog post might be helpful:</p><p><a href="https://medium.com/jim-fleming/loading-a-tensorflow-graph-with-the-c-api-4caaff88463f" rel="nofollow noopener" title="https://medium.com/jim-fleming/loading-a-tensorflow-graph-with-the-c-api-4caaff88463f">https://medium.com/jim-flem...</a></p>

**Comment by joe-antognini** [comment]
Comment written by **Xinxin Zhang** on **09/03/2017 08:44:10**

<p>Hi all, I have a question: I use visual studio 2015, and with a same setting of cmake as you guys use, but every time I built tensorflow, when it comes to the example_trainer project, this or that xxx.obj will be missing, and every time the missed files are different?! eventually I found that I have to for the tf_core_kernels project only to do a rebuild. and that will compile a lot of those missing *.obj files, I wonder what is going on here?</p>

**Comment by joe-antognini** [comment]
Comment written by **Kice** on **09/18/2017 23:59:45**

<p>I would like to know how to add GPU support?</p>

**Comment by joe-antognini** [comment]
Comment written by **Hemecha** on **09/20/2017 08:46:53**

<p>Hi Antognini,</p><p>I have built the static library for c++ and successfully used it in visual studio project (thanks to your tutorial). But, I am trying to build an app in Qt creator, however I have the error Non-OK-status, because I haven´t be able to add the flag "(/wholearchive)" in QtCreator. Could you tell me if it is possible to avoid somehow the flag "/Wholearchive" using another configuration to build the library?</p><p>Thanks in advance.</p>

**Comment by joe-antognini** [comment]
Comment written by **antognini** on **09/20/2017 17:10:00**

<p>I'm not really sure, I'm not familiar with Qt creater... :-/  Sorry I can't be more help here!</p>

**Comment by joe-antognini** [comment]
Comment written by **Hemecha** on **09/21/2017 06:29:43**

<p>That is OK. Thanks anyway.</p>

**Comment by joe-antognini** [comment]
Comment written by **Hemecha** on **09/21/2017 06:33:20**

<p>Hi Alex,</p><p>I can see you´ve built the .dll for tensorflow . Is there any chance that you could tell me if the flag /wholearchive is still needed using the .dll instead of the static library?</p><p>Regards.</p>

**Comment by joe-antognini** [comment]
Comment written by **Hemecha** on **09/21/2017 09:28:50**

<p>Hi again,<br>I have been able to built the .dll tensorflow, but at the end I didn´t need the libprotobuf.dll. I added as dependencies libprotobuF.lib and tensorflow.lib.</p><p>Thanks for your comments and Antognini´s they have been very helpful.</p><p>Kind regards.</p>

**Comment by joe-antognini** [comment]
Comment written by **Benedict Hartmann** on **09/27/2017 17:59:41**

<p>Not being quite sure, but i guess the QT-Creator can handle CMakeLists.txt. On my github account i just rebuilt this example with cmake-support. Feel free to try it: <a href="https://github.com/PinkySan/tensorflowTest" rel="nofollow noopener" title="https://github.com/PinkySan/tensorflowTest">https://github.com/PinkySan...</a></p><p>Please let me know, if it works</p>

**Comment by joe-antognini** [comment]
Comment written by **Manuel Cuevas** on **10/30/2017 16:04:20**

<p>Joe,<br>you mention that it is easier to load a graph already done in python to this C++ API, do you know how to load a graph and values(meta file) created by Python using this Tensorflow C++ API? Can you show an example?</p>

**Comment by joe-antognini** [comment]
Comment written by **Manuel Cuevas** on **11/03/2017 22:10:41**

<p>I am getting a lot of unresolved external symbol for tf_core_lib.lib, tf.cc.lib and tf_corekernels.lib (see image) any idea? <a href="https://uploads.disquscdn.com/images/858bea0a41f354683fcc3d2d5ba33467cb8a783fc979492371263dd4b4ee9649.png" rel="nofollow noopener" title="https://uploads.disquscdn.com/images/858bea0a41f354683fcc3d2d5ba33467cb8a783fc979492371263dd4b4ee9649.png">https://uploads.disquscdn.c...</a></p>

**Comment by joe-antognini** [comment]
Comment written by **Manuel Cuevas** on **11/06/2017 19:35:08**

<p>That symbol is defined in tf_cc_while_loop.lib.<br><a href="https://github.com/tensorflow/tensorflow/issues/14293" rel="nofollow noopener" title="https://github.com/tensorflow/tensorflow/issues/14293">https://github.com/tensorfl...</a></p>

**Comment by joe-antognini** [comment]
Comment written by **Zhang Tianbao** on **11/15/2017 18:29:42**

<p>Hi Joe, thank you very much for this tutorial!</p><p>I was able to follow until "Additional Command Line Options", i.e. the compilation was passed, and "Non-OK-status" error will occur when run the executable. However, when I add those additional command line options, I got tones of linker errors(so much that I just pasted the first three lines):</p><p>1&gt;tf_cc.lib(while_gradients.obj) : error LNK2019: unresolved external symbol "class tensorflow::Status __cdecl tensorflow::ops::BuildWhileLoop(class tensorflow::Scope const &amp;,class std::vector&lt;class tensorflow::output,class="" std::allocator&lt;class="" tensorflow::output=""&gt; &gt; const &amp;,class std::function&lt;class tensorflow::status="" __cdecl(class="" tensorflow::scope="" const="" &amp;,class="" std::vector&lt;class="" tensorflow::output,class="" std::allocator&lt;class="" tensorflow::output=""&gt; &gt; const &amp;,class tensorflow::Output *)&gt; const &amp;,class std::function&lt;class tensorflow::status="" __cdecl(class="" tensorflow::scope="" const="" &amp;,class="" std::vector&lt;class="" tensorflow::output,class="" std::allocator&lt;class="" tensorflow::output=""&gt; &gt; const &amp;,class std::vector&lt;class tensorflow::output,class="" std::allocator&lt;class="" tensorflow::output=""&gt; &gt; *)&gt; const &amp;,class std::basic_string&lt;char,struct std::char_traits&lt;char=""&gt;,class std::allocator&lt;char&gt; &gt; const &amp;,class std::vector&lt;class tensorflow::output,class="" std::allocator&lt;class="" tensorflow::output=""&gt; &gt; *,bool,class tensorflow::Output *)" (?BuildWhileLoop@ops@tensorflow@@YA?AVStatus@2@AEBVScope@2@AEBV?$vector@VOutput@tensorflow@@V?$allocator@VOutput@tensorflow@@@std@@@std@@AEBV?$function@$$A6A?AVStatus@tensorflow@@AEBVScope@2@AEBV?$vector@VOutput@tensorflow@@V?$allocator@VOutput@tensorflow@@@std@@@std@@PEAVOutput@2@@Z@6@AEBV?$function@$$A6A?AVStatus@tensorflow@@AEBVScope@2@AEBV?$vector@VOutput@tensorflow@@V?$allocator@VOutput@tensorflow@@@std@@@std@@PEAV45@@Z@6@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@6@PEAV56@_NPEAVOutput@2@@Z) referenced in function "class tensorflow::Status __cdecl tensorflow::`anonymous namespace'::AddBackPropLoopCounter(class tensorflow::WhileContext *,class tensorflow::Output const &amp;,class tensorflow::Scope const &amp;,class tensorflow::Output *)" (?AddBackPropLoopCounter@?A0x347a03b0@tensorflow@@YA?AVStatus@2@PEAVWhileContext@2@AEBVOutput@2@AEBVScope@2@PEAV52@@Z)<br>1&gt;tf_core_cpu.lib(utils.cc.obj) : error LNK2019: unresolved external symbol cudaGetDeviceProperties referenced in function "class tensorflow::DeviceProperties __cdecl tensorflow::grappler::GetLocalGPUInfo(int)" (?GetLocalGPUInfo@grappler@tensorflow@@YA?AVDeviceProperties@2@H@Z)<br>1&gt;tf_core_cpu.lib(gpu_device.obj) : error LNK2001: unresolved external symbol cudaGetDeviceProperties</p><p>Did you happen to encounter similar problems, or do you have a clue how to solve them? Thank you!</p><p>Kind regards,<br>Tianbao</p>

**Comment by joe-antognini** [comment]
Comment written by **Arnaldo Gualberto** on **11/20/2017 12:41:59**

<p>Congratulations for the tutorial!</p><p>I had to do some additional configurations in order to build the example. I'll let it here just in case to help someonse else who faced the same problems.</p><p>First, I had to open the solution tensorflow.sln present in tensorflow\tensorflow\contrib\cmake\build and build the following projects in Release x64 configurations:<br>- tf_cc<br>- tf_cc_while_loop</p><p>Then, I also added some additional dependencies and includes/library directories:</p><p>Additional Includes Directories:<br>- C:\tensorflow\tensorflow\contrib\cmake\build\nsync\src\nsync\public</p><p>Additional Library Directories<br>- C:\tensorflow\tensorflow\contrib\cmake\build\nsync\src\nsync\Release<br>- C:\tensorflow\tensorflow\contrib\cmake\build\snappy\src\snappy\Release<br>- C:\tensorflow\tensorflow\contrib\cmake\build\sqlite\src\sqlite-build\Release<br>- C:\tensorflow\tensorflow\contrib\cmake\build\tf_cc_while_loop.dir\Release</p><p>Additional Dependencies<br>- nsync.lib<br>- snappy.lib<br>- sqlite.lib<br>- tf_cc_while_loop.lib</p><p>That's it! It's working now!</p>

**Comment by joe-antognini** [comment]
Comment written by **spiovesan** on **11/25/2017 07:27:46**

<p>Great tutorial! I was stuck on " Op type not registered" when including the tf libraries instead of a lot of tf .obj files. For this example  to link, I had to add the two more libraries as Arnaldo Gualberto says and a sqlite3.obj. <br>To avoid "out of memory error" and "Unexpected PDB error" during linking, if using Optimize for debugging (/DEBUG) option, do msbuild in a command prompt (but it ends with a 2.5 gb pdb file using the static libs).</p>

**Comment by joe-antognini** [comment]
Comment written by **Arnaldo Gualberto** on **11/26/2017 10:50:08**

<p>Hi, guys<br>Does someone could build tensorflow for Windows using /MT option?<br>I’m stuck on that and I need some help, please.</p>

**Comment by joe-antognini** [comment]
Comment written by **Thang Duong** on **12/06/2017 18:25:09**

<p>thanks for the comment.  this saved me some time!</p>

**Comment by joe-antognini** [comment]
Comment written by **Arnaldo Gualberto** on **12/06/2017 20:18:07**

<p>You’re welcome!</p>

**Comment by joe-antognini** [comment]
Comment written by **unknownvirus** on **12/11/2017 15:25:55**

<p>Thank you very much for writing this! There area some stuffs needed to be updated like the dependency on tf_cc_while_loop.lib but overall this already really helped me! Finally I was able to run my pb file from tensorflow cpp api...</p><p>I am going to write some tutorial that is related with building tensorflow cpp api in windows, I hope you don't mind if I refer to your blog :)</p><p>Cheers! :D</p>

**Comment by joe-antognini** [comment]
Comment written by **antognini** on **12/11/2017 16:22:27**

<p>It would be great if you could write an updated tutorial!  I'm at a new job now and no longer have access to Windows, so I can't keep this tutorial up to date, and as you can see, there are clearly a lot of people who have questions that I can't answer anymore!</p>

**Comment by joe-antognini** [comment]
Comment written by **firdauslubis88** on **12/12/2017 15:57:57**

<p>Yup, and I already wrote the tutorial here (if you don't mind me to put a URL here):</p><p><a href="http://anonimousindonesian.blogspot.jp/2017/12/tutorial-how-to-build-tensorflow.html" rel="nofollow noopener" title="http://anonimousindonesian.blogspot.jp/2017/12/tutorial-how-to-build-tensorflow.html">http://anonimousindonesian....</a></p><p>I still add link to this site though as additional reference :)</p><p>Thank you. Good luck for your job and keep sharing! :D</p>

**Comment by joe-antognini** [comment]
Comment written by **marzieh mhd** on **01/17/2018 10:50:57**

<p>can you share your .dll library  with us please ??:) if you are please send it to me thank youuu so much: marzieh869@gmail.com</p>

**Comment by joe-antognini** [comment]
Comment written by **Tejaswini teju** on **01/24/2018 05:39:53**

<p>Thanks for the comment @Arnaldo !! it helped me</p>

**Comment by joe-antognini** [comment]
Comment written by **Connor Fitzsimmons** on **02/02/2018 07:19:59**

<p>Hey, i have been having trouble, it seems tf_stream_executor.lib doesnt exist, i have rebuilt from source again and it doesnt seem to show up. If i remove the line from the command line it builds but makes an object? sorry im a bit new to this</p>

**Comment by joe-antognini** [comment]
Comment written by **EKTA PRASHNANI** on **03/06/2018 03:12:55**

<p>Thank you for the tutorial! <br>When I do the steps you mentioned above, i get the correct output; however, I get the following warning: "your cpu supports instructions that this tensorflow binary was not compiled to use avx avx2."</p><p>Could you please specify how to compile tensorflow to use avx, avx2?</p>

**Comment by joe-antognini** [comment]
Comment written by **Jeremy Tavrisov** on **03/19/2018 13:39:43**

<p>Anyone had any success with building in debug while linking to tensorflow.dll?</p>

**Comment by joe-antognini** [comment]
Comment written by **Chubasik** on **04/09/2018 12:39:42**

<p>I had to add some extra above these(Tensorflow 1.7)<br>Additional Libraries:<br>-$(TENSORFLOW_BUILD)\re2\src\re2\Release<br>Additional Dependencies:<br>re2.lib</p>

**Comment by joe-antognini** [comment]
Comment written by **Jerry Louis** on **04/10/2018 15:52:35**

<p>Hi Hemecha,</p><p>I'm trying to do the same as you're, did you finally manage to get it to work with QT? Would it be possible to share what you did to get this to work please? I appreicate your help here.</p>

**Comment by joe-antognini** [comment]
Comment written by **Yuhwan** on **04/18/2018 11:45:47**

<p>Hi there is a error which is fatal error :LNK1120 , How can i fix it....please help me</p><p>Thank you.</p>

**Comment by joe-antognini** [comment]
Comment written by **Yuhwan** on **04/20/2018 13:07:57**

<p>Hi. I have some problems. when I didn't add the additional commands line options. then the build it. the result is " Op type not registered ‘NoOp’ in binary running on users-PC" so i added the command line options. but the errors are even worse that i tried the build without add the command line options. there are 420 errors. that errors are just about LNK1120, LNK2001. can you help me how to fix this??? (I use Tensorflow 1.7 version and CUDA 9.0)</p>

**Comment by joe-antognini** [comment]
Comment written by **Arnaldo Gualberto** on **04/20/2018 13:45:02**

<p>For those who are interested, I wrote a post on Medium and created a Github Repo with detailed instruction about how to build tensorflow on Windows with /MT option (static library).</p><p><a href="https://medium.com/@arnaldog12/how-to-build-tensorflow-on-windows-with-mt-42a8e4bea7e7" rel="nofollow noopener" title="https://medium.com/@arnaldog12/how-to-build-tensorflow-on-windows-with-mt-42a8e4bea7e7">https://medium.com/@arnaldo...</a><br><a href="https://github.com/arnaldog12/Tensorflow-Build-MT" rel="nofollow noopener" title="https://github.com/arnaldog12/Tensorflow-Build-MT">https://github.com/arnaldog...</a></p><p>I hope it helps!</p>

**Comment by joe-antognini** [comment]
Comment written by **Yuhwan** on **04/25/2018 02:13:34**

<p>I found some good thing. I tried the tensorflow 1.7 c++ API. <br>and did it.<br>Here is new things, I follow upon method except the command line options and some library<br>(I added the re2.lib, sqlite.lib, snappy.lib, tf_cc_while_loop.lib)</p><p>the command line options should be<br>/machine:x64 <br>/ignore:4049 /ignore:4197 /ignore:4217 /ignore:4221<br>/WHOLEARCHIVE:tf_cc.lib <br>/WHOLEARCHIVE:tf_cc_framework.lib<br>/WHOLEARCHIVE:tf_cc_ops.lib <br>/WHOLEARCHIVE:tf_core_cpu.lib<br>/WHOLEARCHIVE:tf_core_direct_session.lib <br>/WHOLEARCHIVE:tf_core_framework.lib<br>/WHOLEARCHIVE:tf_core_lib.lib<br>/WHOLEARCHIVE:tf_core_ops.lib   <br>/WHOLEARCHIVE:tf_stream_executor.lib<br>/WHOLEARCHIVE:libjpeg.lib <br>/WHOLEARCHIVE:tf_core_kernels.lib       &lt;------  must display in last line</p><p>you should copy the command line options just like me if you need to use tensorflow 1.7 c++ api</p>

**Comment by joe-antognini** [comment]
Comment written by **ӿøῇχէ** on **10/31/2018 10:50:05**

<p>Don't if anyone will see this...<br>I'm basically trying to run the code you gave here, after making sure that all the "include" and "lib" paths are correct, and I'm getting a ton (91, to be more specific) compilation errors, all of them being some variation of:</p><p></p><pre><code><br>libprotobuf.lib(message_lite.obj) : error LNK2005: "public: bool __cdecl google::protobuf::MessageLite::AppendToString(class std::basic_string&lt;char,struct std::char_traits&lt;char=""&gt;,class std::allocator&lt;char&gt; &gt; *)const " (?AppendToString@MessageLite@protobuf@google@@QEBA_NPEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) already defined in libprotobuf.lib(libprotobuf.dll)<br>libprotobuf.lib(message.obj) : error LNK2005: "public: virtual unsigned __int64 __cdecl google::protobuf::Message::ByteSizeLong(void)const " (?ByteSizeLong@Message@protobuf@google@@UEBA_KXZ) already defined in libprotobuf.lib(libprotobuf.dll)<br></code><br></pre><p></p><p> <a href="https://uploads.disquscdn.com/images/766889fffd3128b0e239b3a7006d2805e4aef4589ce655acdb27172cbeda2a2d.png" rel="nofollow noopener" title="https://uploads.disquscdn.com/images/766889fffd3128b0e239b3a7006d2805e4aef4589ce655acdb27172cbeda2a2d.png">https://uploads.disquscdn.c...</a></p><p>and pointing to various .obj files belonging to libprotobuf.lib.</p>

**Comment by joe-antognini** [comment]
Comment written by **Djalma Simões** on **11/05/2018 12:20:30**

<p>Great article. It worked! Thanks!</p>

**Comment by joe-antognini** [comment]
Comment written by **Prayook Jatesiktat** on **02/09/2019 12:47:34**

<p>Only "/WHOLEARCHIVE:tf_core_kernels.lib" will add almost 5 minutes to the link time.<br>Is there any alternative to this option? I'm on version r1.10 on VS2017 with v140 build tools (GPU).</p>

**Comment by joe-antognini** [comment]
Comment written by **Toshikazu Kuroda** on **06/18/2019 00:00:54**

<p>I know this is a bit old post but I am wondering how you found what objects in the library need to be explicitly linked with the /WHOLEARCHIVE option. I am getting a similar error like,Non-OK-status: sess-&gt;Create(graph_def.graph_def()) status: Not found: Op type not registered 'DivNoNan' in binary running on XXXX. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) tf.contrib.resampler should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.</p><p>I would appreciate if you can give me some advice.</p>

**Comment by joe-antognini** [comment]
Comment written by **Dhaval** on **07/08/2019 20:03:28**

<p>I've been getting the same error. Have you had any luck since?</p>

**Comment by joe-antognini** [comment]
Comment written by **多苍** on **07/15/2019 08:52:24**

<p>do you get any way to fix this?</p>

**Comment by joe-antognini** [comment]
Comment written by **Arnaldo Gualberto** on **09/03/2019 13:19:43**

<p>$100 dollars per hour</p>

**Comment by joe-antognini** [comment]
Comment written by **Arnaldo Gualberto** on **09/03/2019 20:25:32**

<p>arnaldo.g12@gmail.com</p>
